rio/src/RIO/Prelude/RIO.hs

```diff
@@ -35,3 +52,104 @@ instance MonadUnliftIO (RIO env) where
 instance PrimMonad (RIO env) where
     type PrimState (RIO env) = PrimState IO
     primitive = RIO . ReaderT . const . primitive
+
+-- | Abstraction over how to read from and write to a mutable reference
+--
+-- @since 0.1.3.0
+data SomeRef a
+  = SomeRef !(IO a) !(a -> IO ())
+
+-- | Read from a SomeRef
+--
+-- @since 0.1.3.0
+readSomeRef :: SomeRef a -> IO a
+readSomeRef (SomeRef x _) = x
+
+-- | Write to a SomeRef
+--
+-- @since 0.1.3.0
+writeSomeRef :: SomeRef a -> a -> IO ()
+writeSomeRef (SomeRef _ x) = x
+
+-- | Modify a SomeRef
```

Comment:
Can you add a comment that this is subject to race conditions due to lack of atomic operations?